### ISSUE TITLE
dualize in fourier or spatial domain?

### ISSUE THREAD

**Issue by ebrahimebrahim** [ebrahimebrahim/vsvf-monai#3]
When doing the convolution to compute a velocity field from a momentum field, mermaid uses FFT to make it faster on CPU. Convolution on GPU using pytorch is quite fast to compute and to backprop. Which is better for us to use?

Probably spatial smoothing is the better choice, since it's fair to assume that all training is done on GPU, and since it is simpler. But leaving this open until I'm sure we haven't missed something.

[Here is the relevant code in mermaid](https://github.com/uncbiag/mermaid/blob/17d2015666cd228df25269778ce227bf9837150a/mermaid/custom_pytorch_extensions.py); see especially [this todo note](https://github.com/uncbiag/mermaid/blob/17d2015666cd228df25269778ce227bf9837150a/mermaid/custom_pytorch_extensions.py#L6).

[Here is the placeholder where our implementation is slated to go.](https://github.com/ebrahimebrahim/vsvf-monai/blob/4fcc8e9f6e51b2b761d271a033814ca7d7f63479/vsvf.py#L16-L26).

**Comment by ebrahimebrahim** [comment]
Adding here some notes I took during our group meeting on 2022-01-21:

- Dualizing is often done using kernels of a variety of sizes, including huge ones with size almost comparable to image size. Could that perhaps make direct spatial convolution awkward? Not sure about this.
- Padding mode also becomes very important when the kernel is large. Since DFT has an implicit assumption of periodicity we will have to remember to set the padding mode to "repeat," or whatever it's called
